FSW/controllers/controller.py

```diff
@@ -0,0 +1,168 @@
+        ref_ctrl_states = np.zeros((7,))
+        
+        if pointing_target == "Sun":
+            # Define reference torque as zero
+            ref_torque = np.zeros((3,))
+            
+            # Define reference control states with constant estimated sun direction
+            # quaternion defining sun pointing direction
+            # +z pointing to self.state[13:16]
+            # +x cross product of +z and [0,0,1]
+            # +y completes the right handed coordinate system
+            z_body = -est_world_states[13:16] / np.linalg.norm(est_world_states[13:16])
+            x_body = np.cross(z_body, np.array([0,0,1]))
+            y_body = np.cross(z_body, x_body)
+            Rot_mat = np.vstack((x_body, y_body, z_body)).T
+            q_ref = rotmat2quat(Rot_mat)
+        elif pointing_target == "Nadir":
+            att_states = est_world_states[6:13]
+            # Default values if not sun pointing
+            ref_ctrl_states = np.zeros((19,))
+            ref_torque = np.zeros((3,))
+            
+        if  pointing_mode == "detumble":
+            # Default values if not sun pointing
+            ref_ctrl_states = np.zeros((19,))
+            ref_torque = np.zeros((3,))
+        elif pointing_mode == "spin-stabilized":
+            # Default values if not sun pointing
+            ref_ctrl_states = np.zeros((19,))
+            # reference angular velocity in pointing axis 
+            # ref_ctrl_states[:4]  = q_ref
+            # ref_ctrl_states[4:7] = tgt_ang_vel
+            ref_torque = np.zeros((3,))
+        elif pointing_mode == "3D-stabilized":
+            q_ref = rotmat2quat(Rot_mat)
+            ref_ctrl_states[:4] = q_ref
+            ref_torque = np.zeros((3,))
+        else:
+            raise ValueError(f"Unrecognized pointing mode: {pointing_mode}")
+            
+        return ref_torque, ref_ctrl_states
+    
+    def get_torque(self, date, ref_torque, ref_ctrl_states, est_ctrl_states, Idx):
+        # from the reference and estimated quaternions, get the error quaternion for control
+        if self.controller_algo == "Bcross":
+            mtb_torque = self.Bcrossctr.get_dipole_moment_command(est_ctrl_states, Idx)
+            return mtb_torque, []
+        elif self.controller_algo == "Lyapunov":
+            mtb_torque = self.lyapsunpointctr.get_dipole_moment_command(est_ctrl_states, Idx) 
+            return mtb_torque, []
+        elif self.controller_algo == "BaseSP":
+            mtb_torque = self.basesunpointctr.get_dipole_moment_command(est_ctrl_states, Idx) 
+            return mtb_torque, []
+        elif self.controller_algo == "BaseNP":
+            mtb_torque, rw_torque = self.nadirpointctr.get_dipole_moment_command_and_rw_torque(est_ctrl_states, Idx) 
+            return mtb_torque, rw_torque
+        else: 
+            raise ValueError(f"Unrecognized controller algorithm: {self.controller_algo}")
```

Comment:
Rather than checking for the control algorithm identifier everytime we want torque, I think it makes sense to initialize a class attribute that points to the control algorithm callable. So something like
``` python
# on object initialization, point to the callable without passing args
if alg_id == "Bcross":
    self.my_control_alg  = self.lyapsunpointctr.get_dipole_moment_command
```

Reply:
``` python
# everytime we want the control input
self.my_control_alg(my_args)
```

Reply:
changed. If you can lmk if the latest version seems better